docs/contracts/v4/concepts/06-fees.mdx

```diff
@@ -0,0 +1,140 @@
+
+# Swap - Step by Step
+
+The swap happens in a loop until the specified amount has been completely used or the price limit
+has been reached. In each iteration, the code calculates how much of the tokens can be swapped at
+the current price level.
+
+The swap keeps iterating until either the specified amount is fully used or the square root of
+the price hits the defined limit (`sqrtPriceLimitX96`).
+
+# Fees in Uniswap v3
+In Uniswap v3, there are two types of fees: Liquidity Provider (LP) Fees and Protocol Fees.
+
+## Liquidity Provider (LP) Fees in v3
+Whenever a trade (swap) is executed, the trader pays a fee, which is determined by the Uniswap pool's fee rate. The
+fee is a percentage of the trade amount and is added to the pool's liquidity. This fee compensates liquidity providers
+(LPs) for providing liquidity to the pool. The LP fees are implied within the swap calculation and not explicitly
+shown in the code you provided. The fees for LPs are collected through price movement and accumulated in the
+`feeGrowthGlobal0X128` and `feeGrowthGlobal1X128` variables, which track the global fee growth per unit of
+liquidity in terms of `token0` and `token1` respectively.
+
+## Protocol Fees in v3
+These are different from LP fees. The Uniswap protocol may take a portion of the fees for itself, which is separate
+from the fees paid to liquidity providers. The `protocolFees` struct within the contract keeps track of the
+accumulated protocol fees in `token0` and `token1`.
+
+https://github.com/Uniswap/v3-core/blob/main/contracts/UniswapV3Pool.sol
+```solidity
+  // accumulated protocol fees in token0/token1 units
+    struct ProtocolFees {
+        uint128 token0;
+        uint128 token1;
+    }
+    /// @inheritdoc IUniswapV3PoolState
+    ProtocolFees public override protocolFees;
+```
+## Liquidity Provider (LP) Fees vs Protocol Fees
+In essence, the key difference between the two types of fees is that:
+
+- **LP Fees** are distributed among the liquidity providers according to their share of the pool and are accumulated in the
+`feeGrowthGlobal` variables.
+- **Protocol Fees** are extracted from the LP fees and are meant for the Uniswap protocol's treasury or for other
+- purposes as defined by the governance. They are tracked by the protocolFees variable.
+
+## How Fees are Calculated
+A fee is calculated for each swap iteration. It is determined based on the liquidity and the amount being
+swapped in that iteration.
+
+Each swapping loop iteration starts a new swap step. As part of each iteration following properties are being calculated:
+- `step.feeAmount`: This is the fee calculated in each swap step, and it's dependent on the liquidity and price movement within that step.
+- `state.protocolFee`: This accumulates the total protocol fees during the entire swap.
+- `feeGrowthGlobalX128`: This keeps track of the global fee growth, which will be used for liquidity providers to calculate their earned fees.
```

Comment:
Implementation details might do better at the bottom, in an appendix-style section?